### 문제 01. 리스트에 2를 곱하면 무슨 일이 벌어지는가

카페 메뉴 3개의 가격이 있습니다: 아메리카노 4500원, 라떼 5000원, 콜드브루 5500원.

'모든 가격을 2배로 올린 가격표'를 만들고 싶습니다.

파이썬 리스트에 `* 2`를 하면 우리가 원하는 결과가 나올까요?
리스트와 NumPy 배열에 각각 `* 2`를 해 보고, **결과와 길이**를 비교하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1. np로 np.array(lst)로 np 어래이 생성 
2. np.array * 2
3. new price list 에 값저정
```

In [6]:
import numpy as np

item_list =['아메리카노' , '라떼' , '콜드브루']
price_list = [4500, 5000, 5500]


# ===== 여기에 코드를 작성하세요 =====
price_array = np.array(price_list)
print(f'price_array: {price_array}')

price_list2 = price_array * 2
print(f'price_list2: {price_list2}')

price_array: [4500 5000 5500]
price_list2: [ 9000 10000 11000]


<details>
<summary><b>기대 출력 보기 (문제 01)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
리스트 * 2      : [4500, 5000, 5500, 4500, 5000, 5500]
리스트 * 2 길이 : 6
배열 * 2        : [ 9000 10000 11000]
배열 * 2 길이   : 3
</pre>
</details>

### 문제 02. ndim, shape, size 읽기

같은 카페 데이터를 두 가지 모양으로 준비했습니다.

- `cups` : 메뉴 5개의 하루 판매 잔 수 (1차원)
- `sold` : 지점 3곳 × 시간대 4개의 판매 잔 수 (2차원)

두 배열의 `ndim`, `shape`, `size`를 각각 출력하세요.
그리고 `shape`와 `size`가 어떻게 다른지 한 문장으로 설명하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1. 프뢴트 f 기능으로 `ndim`, `shape`, `size`를 각각 출력
2.
3.
```

In [14]:
import numpy as np

cups = np.array([320, 180, 145, 60, 95])

sold = np.array([
    [42,  88, 61, 25],
    [30,  75, 94, 48],
    [55, 102, 47, 12],
])

# ===== 여기에 코드를 작성하세요 =====
print(f'cups data: {cups} \n ndim:{cups.ndim} shape:{cups.shape} size:{cups.size} dtype:{cups.dtype}' )
print(f'sold data: {sold} \n ndim:{sold.ndim} shape:{sold.shape} size:{sold.size} dtype:{sold.dtype}' )

cups data: [320 180 145  60  95] 
 ndim:1 shape:(5,) size:5 dtype:int64
sold data: [[ 42  88  61  25]
 [ 30  75  94  48]
 [ 55 102  47  12]] 
 ndim:2 shape:(3, 4) size:12 dtype:int64


<details>
<summary><b>기대 출력 보기 (문제 02)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
cups
  ndim : 1
  shape: (5,)
  size : 5
sold
  ndim : 2
  shape: (3, 4)
  size : 12
</pre>
</details>

---

## Part 1. 배열은 모양과 데이터 타입을 가진다

배열은 `ndim`(차원 수), `shape`(모양), `size`(값 개수), `dtype`(데이터 종류)를 가집니다.
정확한 비트 수보다 **정수인지 실수인지 문자열인지**를 먼저 봅니다.

### 문제 03. dtype은 하나로 통일된다

아래 세 배열을 만들고 각각의 값과 `dtype`을 출력하세요.

1. `[10, 20, 30]`
2. `[10, 20, 30.5]`
3. `[10, 20, "30"]`

출력 결과를 보고, NumPy가 서로 다른 타입을 만났을 때 **어느 쪽으로 통일하는지** 한 문장으로 적으세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [17]:
import numpy as np


# ===== 여기에 코드를 작성하세요 =====
arr_data1 = np.array([10, 20, 30])
arr_data2 = np.array([10, 20, 30.5])
arr_data3 = np.array([10, 20, "30"])

print(f'arr_data1: {arr_data1} \n ndim:{arr_data1.ndim} shape:{arr_data1.shape} size:{arr_data1.size} dtype:{arr_data1.dtype}' )
print(f'arr_data2: {arr_data2} \n ndim:{arr_data2.ndim} shape:{arr_data2.shape} size:{arr_data2.size} dtype:{arr_data2.dtype}' )
print(f'arr_data3: {arr_data2} \n ndim:{arr_data3.ndim} shape:{arr_data3.shape} size:{arr_data3.size} dtype:{arr_data3.dtype}' )


arr_data1: [10 20 30] 
 ndim:1 shape:(3,) size:3 dtype:int64
arr_data2: [10.  20.  30.5] 
 ndim:1 shape:(3,) size:3 dtype:float64
arr_data3: [10.  20.  30.5] 
 ndim:1 shape:(3,) size:3 dtype:<U21


<details>
<summary><b>기대 출력 보기 (문제 03)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
[10 20 30] -&gt; int64
[10.  20.  30.5] -&gt; float64
[&#x27;10&#x27; &#x27;20&#x27; &#x27;30&#x27;] -&gt; &lt;U21
</pre>
</details>

### 문제 04. astype으로 타입 바꾸기 (버림에 주의)

매장 4곳의 실내 온도가 실수로 기록되어 있습니다.

```python
temps = np.array([21.7, 24.3, 19.8, 26.5])
```

이 값을 정수형으로 바꾼 배열을 만들고, 변환 전/후의 값과 dtype을 출력하세요.
그리고 **원본 `temps`가 그대로 남아 있는지**도 함께 출력해 확인하세요.

변환 결과를 보고 '반올림인지 버림인지' 판단해 적으세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [21]:
import numpy as np

temps = np.array([21.7, 24.3, 19.8, 26.5])

# ===== 여기에 코드를 작성하세요 =====
print(f'temps.dtype: {temps.dtype} ')
print(f'temps data: {temps}')

temps_int = temps.astype(int) 
print(f'temps_int.dtype afer temps.astype(int) : {temps_int.dtype} ')
print(f'temps_int data: {temps_int}')

temps.dtype: float64 
temps data: [21.7 24.3 19.8 26.5]
temps_int.dtype afer temps.astype(int) : int64 
temps_int data: [21 24 19 26]


<details>
<summary><b>기대 출력 보기 (문제 04)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
변환 전: [21.7 24.3 19.8 26.5] float64
변환 후: [21 24 19 26] int64
원본 유지: [21.7 24.3 19.8 26.5]
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Part 2. NumPy가 배열을 대신 만들어 준다

`np.arange(start, stop, step)` / `np.zeros(shape)` / `np.ones(shape)`

반복 앵커: **시작은 포함, 끝은 제외**

### 문제 05. arange로 규칙적인 수열 만들기

`np.arange`로 다음 세 배열을 만들고, 각각 값과 `size`를 출력하세요.

1. 0부터 9까지 (10개)
2. 6부터 20까지 (1씩 증가)
3. 5부터 50까지 5씩 증가

**핵심 앵커: 시작은 포함, 끝은 제외.** stop에 무엇을 넣어야 하는지 먼저 종이에 적으세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1. import mumpy as np 선언
2. np.arange() 함수 이용
3.
```

In [24]:
import numpy as np

# ===== 여기에 코드를 작성하세요 =====
arr1 = np.arange(0, 10)
print(f'arr1 data: {arr1}')

arr2 = np.arange(6, 21, 1)
print(f'arr2 data: {arr2}')

arr3 = np.arange(5, 51, 5)
print(f'arr3 data: {arr3}')



arr1 data: [0 1 2 3 4 5 6 7 8 9]
arr2 data: [ 6  7  8  9 10 11 12 13 14 15 16 17 18 19 20]
arr3 data: [ 5 10 15 20 25 30 35 40 45 50]


<details>
<summary><b>기대 출력 보기 (문제 05)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
seq_a: [0 1 2 3 4 5 6 7 8 9] | size = 10
seq_b: [ 6  7  8  9 10 11 12 13 14 15 16 17 18 19 20] | size = 15
seq_c: [ 5 10 15 20 25 30 35 40 45 50] | size = 10
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 06. zeros와 ones로 빈 표 만들기

4주 동안 3개 지점의 예약 건수를 기록할 **빈 표**를 미리 만들어 두려고 합니다.

1. 4행 3열, 값이 전부 정수 `0`인 표를 만드세요.
2. 2행 5열, 값이 전부 `1.0`인 표를 만드세요.
3. 두 표의 `shape`와 `dtype`을 출력하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [29]:
import numpy as np

# ===== 여기에 코드를 작성하세요 ===== np.zeros((2, 3), dtype=int) / np.zeros((2, 3), dtype=float) / np.ones((2, 3), dtype=float)
reservation1 = np.zeros((4, 3), dtype=int)
print(f'reservation1 4*3 : {reservation1} \n shape: {reservation1.shape} dtype: {reservation1.dtype}')

reservation2= np.ones((2, 5))
print(f'reservation2 2*5 : {reservation2} \n shape: {reservation2.shape} dtype: {reservation2.dtype}')

reservation1 4*3 : [[0 0 0]
 [0 0 0]
 [0 0 0]
 [0 0 0]] 
 shape: (4, 3) dtype: int64
reservation2 2*5 : [[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]] 
 shape: (2, 5) dtype: float64


<details>
<summary><b>기대 출력 보기 (문제 06)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
예약 초기표:
 [[0 0 0]
 [0 0 0]
 [0 0 0]
 [0 0 0]]
1로 채운 표:
 [[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]]
shape: (4, 3) (2, 5)
dtype: int64 float64
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Part 3. 배열의 모양 바꾸기

`reshape`는 값의 순서를 유지하면서 모양만 바꿉니다.

반복 질문: **숫자가 총 몇 개인가?**

### 문제 07. reshape로 같은 값의 모양만 바꾸기

`np.arange(12)`로 값 12개를 만든 뒤, 다음 세 가지 모양으로 각각 바꾸세요.

1. 3행 4열
2. 4행 3열
3. 2행 6열

각 결과와 `shape`, `size`를 출력하고, **원본 배열이 바뀌지 않았는지**도 확인하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1. np 인스턴스의 메소드 reshape 이용
2. size가 배열의 갯수와 같아야 한다
3.
```

In [32]:
import numpy as np

base = np.arange(12)

# ===== 여기에 코드를 작성하세요 =====

three_by_four = base.reshape(3, 4)
four_by_three = base.reshape(4, 3)
two_by_six = base.reshape(2, 6)

print("원본:", base, "shape=", base.shape)
print("3행 4열:\n", three_by_four)
print("4행 3열:\n", four_by_three)
print("2행 6열:\n", two_by_six)
print("size:", base.size, three_by_four.size, four_by_three.size, two_by_six.size)

원본: [ 0  1  2  3  4  5  6  7  8  9 10 11] shape= (12,)
3행 4열:
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
4행 3열:
 [[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]
2행 6열:
 [[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]]
size: 12 12 12 12


<details>
<summary><b>기대 출력 보기 (문제 07)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
원본: [ 0  1  2  3  4  5  6  7  8  9 10 11] | shape = (12,)
3행 4열:
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
4행 3열:
 [[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]
2행 6열:
 [[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]]
size 비교: 12 12 12 12
</pre>
</details>

### 문제 08. reshape가 실패하는 이유를 직접 확인하고 고치기

값 12개를 `reshape(5, 3)`으로 바꾸려고 하면 어떻게 될까요?

**먼저 예측하세요.** 성공할까요, 실패할까요? 실패한다면 왜일까요?

그다음 `try / except ValueError`로 감싸서 실행하고, 오류의 **종류 이름**과 **이유**를 출력하세요.
마지막으로 값 개수를 바꾸지 않고 성공하는 모양 하나를 골라 다시 실행하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [39]:
import numpy as np

base = np.arange(12)


# ===== 여기에 코드를 작성하세요 =====
try:
    five_by_three = base.reshape(5, 3)

except ValueError as error:
    print("예상한 오류 종류:", type(error).__name__)
    print("이유: 값 12개를 15칸 모양에 넣을 수 없습니다.")

else:
    print("5행 3열:\n", five_by_three )

retry = base.reshape(6, 2)
print(retry)
print("retry shape:", retry.shape, "size:", retry.size)

예상한 오류 종류: ValueError
이유: 값 12개를 15칸 모양에 넣을 수 없습니다.
[[ 0  1]
 [ 2  3]
 [ 4  5]
 [ 6  7]
 [ 8  9]
 [10 11]]
retry shape: (6, 2) size: 12


<details>
<summary><b>기대 출력 보기 (문제 08)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
예상한 오류 종류: ValueError
이유: 값 12개를 15 칸 모양에 넣을 수 없습니다.
원본 size: 12
수정 후:
 [[ 0  1]
 [ 2  3]
 [ 4  5]
 [ 6  7]
 [ 8  9]
 [10 11]]
retry shape: (6, 2) | size: 12
</pre>
</details>

### 문제 09. -1을 써서 나머지 축을 자동으로 계산하게 하기

값 24개를 만든 뒤, 다음 두 가지를 `-1`을 사용해서 만드세요.

1. 열이 6개이고 행 개수는 NumPy가 알아서 정하게 하기
2. 행이 4개이고 열 개수는 NumPy가 알아서 정하게 하기

각각의 `shape`를 출력해서 몇 행 몇 열이 되었는지 확인하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [41]:
import numpy as np

base24 = np.arange(24)


# ===== 여기에 코드를 작성하세요 =====
arr_x6 = base24.reshape(-1, 6)
print(f'reshape(-1, 6) shape ->: {arr_x6.shape}')
print(arr_x6)

arr_4x = base24.reshape(4, -1)
print(f'reshape(4, -1) shape ->: {arr_4x.shape}')
print(arr_4x)


reshape(-1, 6) shape ->: (4, 6)
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]
reshape(4, -1) shape ->: (4, 6)
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]


<details>
<summary><b>기대 출력 보기 (문제 09)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
reshape(-1, 6) shape -&gt; (4, 6)
reshape(4, -1) shape -&gt; (4, 6)
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Part 4. 원하는 데이터를 위치로 찾기 (1차원)

- 첫 번째 위치: `0`
- 마지막 위치: `-1`
- 슬라이싱 `a[start:stop]`은 **start 포함, stop 제외**

### 문제 10. 위치 번호로 값 하나 꺼내기

카페의 1월~12월 월별 방문자 수 데이터가 있습니다.

`months`와 `visitors`를 만든 뒤 다음을 출력하세요.

1. 1월 방문자 수
2. 7월 방문자 수
3. 마지막 달(12월) 방문자 수 — **음수 인덱스를 사용**
4. 뒤에서 두 번째 달 방문자 수

`months`와 `visitors`의 길이가 같은지도 먼저 확인하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [48]:
import numpy as np

months = np.arange(1, 13)
visitors = np.array([1820, 1650, 2010, 2240, 2380, 2510,
                     2760, 2890, 2430, 2180, 1970, 2300])


# ===== 여기에 코드를 작성하세요 =====
print("같은 길이인가?", len(months) == len(visitors))
print(f'months : {months}')
print(f'visitors : {visitors}')

print(f'1월 방문자수 : {visitors[0]}')
print(f'7월 방문자수 : {visitors[6]}')
print(f'12월 방문자수 : {visitors[-1]}')
print(f'뒤에서 두 번째 달 방문자수 : {visitors[-2]}')



같은 길이인가? True
months : [ 1  2  3  4  5  6  7  8  9 10 11 12]
visitors : [1820 1650 2010 2240 2380 2510 2760 2890 2430 2180 1970 2300]
1월 방문자수 : 1820
7월 방문자수 : 2760
12월 방문자수 : 2300
뒤에서 두 번째 달 방문자수 : 1970


<details>
<summary><b>기대 출력 보기 (문제 10)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
월: [ 1  2  3  4  5  6  7  8  9 10 11 12]
방문자: [1820 1650 2010 2240 2380 2510 2760 2890 2430 2180 1970 2300]
길이가 같은가? True
1월 방문자      : 1820
7월 방문자      : 2760
마지막 달 방문자: 2300
뒤에서 두 번째  : 1970
</pre>
</details>

### 문제 11. 슬라이싱으로 구간 꺼내기

같은 `visitors` 배열에서 다음 네 가지를 출력하세요.

1. 3월부터 5월까지 (사람이 세는 기준)
2. 처음 4개월
3. 9월부터 끝까지
4. 1월부터 두 달 간격으로

각각 코드를 쓰기 전에 `start`와 `stop`을 종이에 먼저 적으세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3. `start`와 `stop` [0::2]
```

In [55]:
import numpy as np

visitors = np.array([1820, 1650, 2010, 2240, 2380, 2510,
                     2760, 2890, 2430, 2180, 1970, 2300])


# ===== 여기에 코드를 작성하세요 =====
print(f'3월부터 5월까지: {visitors[2:5]}')
print(f'처음 4개월: {visitors[0:4]}')
print(f'9월부터 끝까지: {visitors[8:]}')
print(f'1월부터 두 달 간격으로: {visitors[::2]}')


3월부터 5월까지: [2010 2240 2380]
처음 4개월: [1820 1650 2010 2240]
9월부터 끝까지: [2430 2180 1970 2300]
1월부터 두 달 간격으로: [1820 2010 2380 2760 2430 1970]


<details>
<summary><b>기대 출력 보기 (문제 11)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
3~5월    : [2010 2240 2380]
처음 4개월: [1820 1650 2010 2240]
9월~12월 : [2430 2180 1970 2300]
두 달 간격: [1820 2010 2380 2760 2430 1970]
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 12. copy를 쓰지 않으면 원본이 함께 바뀐다

1월 방문자 수가 잘못 입력되어 `1820`을 `2000`으로 고치려 합니다.

두 가지를 비교해 보세요.

1. `visitors.copy()`로 사본을 만들어 첫 값을 바꾼 뒤, **원본**의 첫 값을 출력
2. 슬라이싱 `visitors[:3]`을 변수에 저장한 뒤 그 첫 값을 바꾸고, **원본**의 첫 값을 출력

두 경우에서 원본이 어떻게 달라지는지 관찰하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [58]:
import numpy as np

visitors = np.array([1820, 1650, 2010, 2240, 2380, 2510,
                     2760, 2890, 2430, 2180, 1970, 2300])


# ===== 여기에 코드를 작성하세요 =====
visitors_copy = visitors.copy()
visitors_copy[0] = 2000
print(f'[copy 사용]')
print(f'   사본 첫 값 : {visitors_copy[0]}')
print(f'   원본 첫 값 : {visitors[0]}')

visitors_slice = visitors[:3]
visitors_slice[0] = 2000
print(f'[슬라이싱 사용]')
print(f'   참조어레이 첫 값 : {visitors_slice[0]}')
print(f'   원본 첫 값 : {visitors[0]}')

[copy 사용]
   사본 첫 값 : 2000
   원본 첫 값 : 1820
[슬라이싱 사용]
   참조어레이 첫 값 : 2000
   원본 첫 값 : 2000


<details>
<summary><b>기대 출력 보기 (문제 12)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
[copy 사용]
  사본 첫 값 : 2000
  원본 첫 값 : 1820 &lt;- 그대로 유지됨
[슬라이싱 사용]
  잘라낸 것 첫 값: 2000
  원본 첫 값     : 2000 &lt;- 같이 바뀜
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Part 5. 표에서 원하는 값 찾기 (2차원)

2차원 배열의 기본 문장은 `array[행, 열]` 입니다.
콜론 하나 `:`는 그 축을 **전부** 선택합니다.

### 문제 13. 표에서 값 한 칸 꺼내기

지점 3곳 × 시간대 4개의 판매 잔 수 표입니다.

```text
        아침  점심  저녁  야간
강남     42    88    61    25
홍대     30    75    94    48
판교     55   102    47    12
```

다음 값을 출력하세요.

1. 홍대 지점의 저녁 판매량
2. 판교 지점의 점심 판매량
3. 강남 지점의 야간 판매량

각각 `[행, 열]`의 숫자를 먼저 종이에 적고 코드로 옮기세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [59]:
import numpy as np

branches = np.array(["강남", "홍대", "판교"])
timeslots = np.array(["아침", "점심", "저녁", "야간"])
sold = np.array([
    [42,  88, 61, 25],
    [30,  75, 94, 48],
    [55, 102, 47, 12],
])


# ===== 여기에 코드를 작성하세요 =====
print(f'1. 홍대 지점의 저녁 판매량: {sold[1, 2]}')
print(f'2. 판교 지점의 점심 판매량: {sold[2, 1]}')
print(f'3. 강남 지점의 야간 판매량: {sold[0, 3]}')

1. 홍대 지점의 저녁 판매량: 94
2. 판교 지점의 점심 판매량: 102
3. 강남 지점의 야간 판매량: 25


<details>
<summary><b>기대 출력 보기 (문제 13)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
표:
 [[ 42  88  61  25]
 [ 30  75  94  48]
 [ 55 102  47  12]]
shape: (3, 4) (행 = 지점 3, 열 = 시간대 4)
홍대 저녁: 94
판교 점심: 102
강남 야간: 25
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 14. 표에서 직사각형 잘라내기

같은 `sold` 표에서 다음 직사각형을 잘라 출력하세요.

- 행: 홍대, 판교
- 열: 점심, 저녁

결과와 `shape`를 함께 출력하고, 결과가 왜 `(2, 2)`인지 설명하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**
        아침  점심  저녁  야간
강남     42    88    61    25
홍대     30    75    94    48
판교     55   102    47    12
```text
1.
2.
3.
```

In [64]:
import numpy as np

sold = np.array([
    [42,  88, 61, 25],
    [30,  75, 94, 48],
    [55, 102, 47, 12],
])

# ===== 여기에 코드를 작성하세요 =====
sold_slice = sold[1:3, 1:3]
print(f'행: 홍대, 판교 * 열: 점심, 저녁: {sold_slice}, shape: {sold_slice.shape}')

행: 홍대, 판교 * 열: 점심, 저녁: [[ 75  94]
 [102  47]], shape: (2, 2)


<details>
<summary><b>기대 출력 보기 (문제 14)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
홍대·판교 × 점심·저녁:
 [[ 75  94]
 [102  47]]
shape: (2, 2)
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 15. 행 전체, 열 전체 선택하기

같은 `sold` 표에서 다음을 출력하세요.

1. 강남 지점의 모든 시간대 판매량
2. 모든 지점의 야간 판매량
3. 모든 지점의 점심~저녁 판매량

각 결과의 `shape`도 같이 출력해서, 1차원이 되는지 2차원이 되는지 확인하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**
        아침  점심  저녁  야간
강남     42    88    61    25
홍대     30    75    94    48
판교     55   102    47    12
```text
1.
2.
3.
```

In [84]:
import numpy as np

sold = np.array([
    [42,  88, 61, 25],
    [30,  75, 94, 48],
    [55, 102, 47, 12],
])


# ===== 여기에 코드를 작성하세요 =====

print(f'1.강남 지점의 모든 시간대 판매량: \n {sold[0:1, :]} / shape: {sold[0:1, :].shape}')
print(f'1.강남 지점의 모든 시간대 판매량: \n {sold[0, :]} / shape: {sold[0, :].shape}')
print(f'2.모든 지점의 야간 판매량: \n {sold[0:, 3:4]} / shape: {sold[:, 3:4].shape}')
print(f'2.모든 지점의 야간 판매량: \n {sold[0:, 3]} / shape: {sold[:, 3].shape}')
print(f'3.모든 지점의 점심~저녁 판매량: \n {sold[0:, 1:3]} / shape: {sold[:, 1:3].shape}')

1.강남 지점의 모든 시간대 판매량: 
 [[42 88 61 25]] / shape: (1, 4)
1.강남 지점의 모든 시간대 판매량: 
 [42 88 61 25] / shape: (4,)
2.모든 지점의 야간 판매량: 
 [[25]
 [48]
 [12]] / shape: (3, 1)
2.모든 지점의 야간 판매량: 
 [25 48 12] / shape: (3,)
3.모든 지점의 점심~저녁 판매량: 
 [[ 88  61]
 [ 75  94]
 [102  47]] / shape: (3, 2)


<details>
<summary><b>기대 출력 보기 (문제 15)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
강남 전체 시간대: [42 88 61 25] | shape = (4,)
모든 지점 야간  : [25 48 12] | shape = (3,)
모든 지점 점심~저녁:
 [[ 88  61]
 [ 75  94]
 [102  47]] 
| shape = (3, 2)
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Part 6. 조건에 맞는 데이터 찾기

실제 분석 질문은 위치를 알려 주지 않습니다.

1. 조건을 실행해 `True/False` 배열을 본다
2. 그 배열을 mask로 저장한다
3. `array[mask]`로 True 위치의 값만 꺼낸다

### 문제 16. 조건 배열(mask)을 만들고 값 꺼내기

월별 방문자 데이터에서 **방문자가 2,400명을 넘은 달**을 찾으려 합니다.

순서대로 진행하세요.

1. `visitors > 2400` 을 그대로 출력해서 무엇이 나오는지 본다
2. 그 결과의 `dtype`과 `shape`를 확인한다
3. mask를 사용해 해당하는 **월**과 **방문자 수**를 각각 꺼낸다
4. 꺼낸 값이 정말 모두 2,400보다 큰지 `np.all()`로 검증한다

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [104]:
import numpy as np

months = np.arange(1, 13)
visitors = np.array([1820, 1650, 2010, 2240, 2380, 2510,
                     2760, 2890, 2430, 2180, 1970, 2300])


# ===== 여기에 코드를 작성하세요 =====
over_2400 = visitors > 2400
print(f'over_2400: {over_2400} / dtype: {over_2400.dtype}')
print(visitors > 2400 )

selected_visitors = visitors[over_2400]
print(selected_visitors)
print("모두 조건 만족?",  np.all(selected_visitors > 2400))

over_2400: [False False False False False  True  True  True  True False False False] / dtype: bool
[False False False False False  True  True  True  True False False False]
[2510 2760 2890 2430]
모두 조건 만족? True


<details>
<summary><b>기대 출력 보기 (문제 16)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
조건 결과: [False False False False False  True  True  True  True False False False]
조건 dtype: bool
원본과 shape가 같은가? True
2,400명 초과한 달 : [6 7 8 9]
그 달의 방문자    : [2510 2760 2890 2430]
모두 조건을 만족하는가? True
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 17. 조건 두 개를 묶기 (& 와 |)

같은 데이터에서 다음 두 가지를 구하세요.

1. 방문자가 **2,000명 이상 2,500명 이하**인 달의 방문자 수
2. 방문자가 **1,900명 미만이거나 2,800명 초과**인 달의 방문자 수

주의: 파이썬의 `and` / `or` 가 아니라 NumPy 전용 기호를 씁니다.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [108]:
import numpy as np

visitors = np.array([1820, 1650, 2010, 2240, 2380, 2510,
                     2760, 2890, 2430, 2180, 1970, 2300])


# ===== 여기에 코드를 작성하세요 =====
print(f'방문자가 2,000명 이상 2,500명 이하인 달의 방문자 수: {visitors[(visitors >= 2000) & (visitors <= 2500)]}')
print(f'방문자가 1,900명 미만이거나 2,800명 초과인 달의 방문자 수: {visitors[(visitors < 1900) | (visitors > 2800)]}')

방문자가 2,000명 이상 2,500명 이하인 달의 방문자 수: [2010 2240 2380 2430 2180 2300]
방문자가 1,900명 미만이거나 2,800명 초과인 달의 방문자 수: [1820 1650 2890]


<details>
<summary><b>기대 출력 보기 (문제 17)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
2,000 이상 2,500 이하: [2010 2240 2380 2430 2180 2300]
1,900 미만 또는 2,800 초과: [1820 1650 2890]
각각 개수: 6 3
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 18. 조건에 맞는 값이 하나도 없을 때

방문자가 **5,000명을 넘은 달**을 찾아보세요. 실제로는 그런 달이 없습니다.

오류가 날까요, 아니면 다른 결과가 나올까요? **먼저 예측한 뒤** 실행하세요.

결과의 값, `shape`, `size`를 출력하고, `size`로 '결과가 비었는지'를 판단하는 코드도 작성하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [110]:
import numpy as np

visitors = np.array([1820, 1650, 2010, 2240, 2380, 2510,
                     2760, 2890, 2430, 2180, 1970, 2300])


# ===== 여기에 코드를 작성하세요 =====
print(f'방문자가 5,000명 넘는 달의 방문자 수: {visitors[visitors > 5000]} shape: {visitors[visitors > 5000].shape} size: {visitors[visitors > 5000].size}')

방문자가 5,000명 넘는 달의 방문자 수: [] shape: (0,) size: 0


<details>
<summary><b>기대 출력 보기 (문제 18)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
결과: []
shape: (0,)
size : 0
해당 조건을 만족하는 달이 없습니다.
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Part 7. 한꺼번에 계산하기

`.sum()` `.mean()` `.min()` `.max()`
배열 × 숫자는 모든 원소에, 같은 shape 배열끼리는 같은 위치끼리 계산됩니다.

### 문제 19. 합계, 평균, 최소, 최대

월별 방문자 데이터로 다음을 출력하세요.

1. 연간 총 방문자 수
2. 월평균 방문자 수 (소수점 첫째 자리까지 반올림)
3. 가장 적었던 달의 방문자 수
4. 가장 많았던 달의 방문자 수
5. **가장 많았던 달이 몇 월인지** (`argmax` 사용)

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [117]:
import numpy as np

months = np.arange(1, 13)
visitors = np.array([1820, 1650, 2010, 2240, 2380, 2510,
                     2760, 2890, 2430, 2180, 1970, 2300])


# ===== 여기에 코드를 작성하세요 =====
print(f'연간 총 방문자 수: {visitors.sum()} ')
print(f'월평균 방문자 수 (소수점 첫째 자리까지 반올림): {round(visitors.mean(),1)} ')
print(f'가장 적었던 달의 방문자 수: {visitors.min()} ')
print(f'가장 많았던 달의 방문자 수: {visitors.max()} ')
print(f'가장 많았던 달의 위치: {visitors.argmax()} ')
print(f'가장 많았던 달이 몇 월인지: {months[visitors.argmax()]}월 ')

연간 총 방문자 수: 27140 
월평균 방문자 수 (소수점 첫째 자리까지 반올림): 2261.7 
가장 적었던 달의 방문자 수: 1650 
가장 많았던 달의 방문자 수: 2890 
가장 많았던 달의 위치: 7 
가장 많았던 달이 몇 월인지: 8월 


<details>
<summary><b>기대 출력 보기 (문제 19)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
연간 총 방문자: 27140
월평균 방문자 : 2261.7
최소 방문자   : 1650
최대 방문자   : 2890
최대인 위치   : 7
최대인 달     : 8 월
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 20. 배열끼리 계산해서 매출 구하기 (처음 질문의 답)

맨 처음 질문으로 돌아갑니다. 메뉴 5개의 가격과 판매 잔 수로 다음을 구하세요.

1. 메뉴별 매출 (가격 × 잔 수)
2. 전체 매출
3. 전체 매출에서 각 메뉴가 차지하는 비율(%) — 소수점 첫째 자리까지
4. 모든 가격을 10% 인상했을 때의 새 가격표

**한 줄로 표현되는지** 확인하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [134]:
import numpy as np

menu = np.array(["아메리카노", "라떼", "콜드브루", "녹차라떼", "레모네이드"])
price = np.array([4500, 5000, 5500, 5300, 6000])
cups = np.array([320, 180, 145, 60, 95])


# ===== 여기에 코드를 작성하세요 =====

print(f'길이 체크 price: {len(price)} cups:{len(cups)}  {len(price) == len(cups)}')
salse_by_menu = price * cups
print(f'메뉴         : {menu}')
print(f'메뉴별 매출   :  {salse_by_menu }')
print(f'전체 매출  :  {salse_by_menu.sum():,}')
print(f'전체 매출에서 각 메뉴가 차지하는 비율(%) — 소수점 첫째 자리까지 :  { np.round( salse_by_menu / salse_by_menu.sum() *100 , 1) }')
price_new = price * (1 + 0.1)
print(f'모든 가격을 10% 인상했을 때의 새 가격표 :  { price_new }')


길이 체크 price: 5 cups:5  True
메뉴         : ['아메리카노' '라떼' '콜드브루' '녹차라떼' '레모네이드']
메뉴별 매출   :  [1440000  900000  797500  318000  570000]
전체 매출  :  4,025,500
전체 매출에서 각 메뉴가 차지하는 비율(%) — 소수점 첫째 자리까지 :  [35.8 22.4 19.8  7.9 14.2]
모든 가격을 10% 인상했을 때의 새 가격표 :  [4950. 5500. 6050. 5830. 6600.]


<details>
<summary><b>기대 출력 보기 (문제 20)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
메뉴      : [&#x27;아메리카노&#x27; &#x27;라떼&#x27; &#x27;콜드브루&#x27; &#x27;녹차라떼&#x27; &#x27;레모네이드&#x27;]
메뉴별 매출: [1440000  900000  797500  318000  570000]
전체 매출  : 4025500
매출 비율(%): [35.8 22.4 19.8  7.9 14.2]
10% 인상가 : [4950. 5500. 6050. 5830. 6600.]
shape 확인 : (5,) (5,) (5,)
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Support. 더 유연하게 고르고 정렬하기

- fancy indexing: 떨어진 위치 여러 개를 한 번에 선택
- `np.sort` / `.sort()` / `argsort`
- `axis=0` 은 열마다 결과 하나, `axis=1` 은 행마다 결과 하나

### 문제 21. fancy indexing - 떨어진 위치 여러 개 고르기

`visitors`에서 **1분기의 마지막 달(3월), 2분기의 마지막 달(6월), 3분기의 마지막 달(9월)**
세 달만 골라 월과 방문자 수를 출력하세요.

슬라이싱으로는 붙어 있는 구간만 자를 수 있습니다. 떨어진 위치는 **인덱스 배열**로 고릅니다.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [145]:
import numpy as np

months = np.arange(1, 13)
visitors = np.array([1820, 1650, 2010, 2240, 2380, 2510,
                     2760, 2890, 2430, 2180, 1970, 2300])


# ===== 여기에 코드를 작성하세요 =====
chosen_indices = np.array([2,5,8])
print(chosen_indices)
print(f'3월 6월 9월 {visitors[chosen_indices]} shape: {visitors[chosen_indices].shape}')
print(f'3월 6월 9월 {visitors[np.array([2,5,8])]} shape: { visitors[ np.array([2,5,8])].shape }')



[2 5 8]
3월 6월 9월 [2010 2510 2430] shape: (3,)
3월 6월 9월 [2010 2510 2430] shape: (3,)


<details>
<summary><b>기대 출력 보기 (문제 21)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
선택한 위치: [2 5 8]
선택한 월   : [3 6 9]
선택한 방문자: [2010 2510 2430]
결과 shape  : (3,)
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 22. np.sort 와 .sort() 의 차이

바리스타 5명의 하루 제조 잔 수가 있습니다.

```python
made = np.array([182, 240, 165, 258, 203])
```

1. `np.sort(made)` 결과와, 그 뒤의 **원본** `made`를 출력하세요.
2. `made`의 사본을 만든 뒤 `.sort()`를 호출하고, **반환값**과 **사본**을 출력하세요.
3. 내림차순으로 정렬한 결과도 출력하세요.

`np.sort`와 `.sort()`가 무엇이 다른지 한 문장으로 정리하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [147]:
import numpy as np

made = np.array([182, 240, 165, 258, 203])


# ===== 여기에 코드를 작성하세요 =====
print(f' made: {made}')
print(f' np.sort(made) : {np.sort(made)}')
print(f' made: {made}')

print(f' made.sort(): {made.sort()}')
print(f' made: {made}')


 made: [182 240 165 258 203]
 np.sort(made) : [165 182 203 240 258]
 made: [182 240 165 258 203]
 made.sort(): None
 made: [165 182 203 240 258]


<details>
<summary><b>기대 출력 보기 (문제 22)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
np.sort 결과: [165 182 203 240 258]
원본 유지   : [182 240 165 258 203]
.sort() 반환값: None
.sort() 후 사본: [165 182 203 240 258]
내림차순: [258 240 203 182 165]
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 23. argsort로 상위 3명 뽑기

바리스타 이름과 제조 잔 수가 짝을 이루고 있습니다.

```python
baristas = np.array(["민준", "서연", "도윤", "하윤", "지호"])
made = np.array([182, 240, 165, 258, 203])
```

제조 잔 수가 **많은 순서로 상위 3명의 이름과 잔 수**를 출력하세요.

`np.sort`로는 이름을 함께 옮길 수 없습니다. **위치를 먼저 정렬**해야 합니다.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [157]:
import numpy as np

baristas = np.array(["민준", "서연", "도윤", "하윤", "지호"])
made = np.array([182, 240, 165, 258, 203])


# [옮기기] argsort는 '정렬된 값'이 아니라 '정렬 순서대로의 위치'를 준다
order_asc = np.argsort(made) #-- 작은값에서 큰값의 순서로 위치 인덱스 번호 반환
print("오름차순 위치:", order_asc)
print("확인 - 그 위치의 값:", made[order_asc])

# [옮기기] 뒤집어서 내림차순 위치로 만들고, 앞 3개만 자른다
top3 = np.argsort(made)[::-1][:3]

print("상위 3명 위치:", top3)
print("상위 3명 이름:", baristas[top3])
print("상위 3명 잔 수:", made[top3])


오름차순 위치: [2 0 4 1 3]
확인 - 그 위치의 값: [165 182 203 240 258]
상위 3명 위치: [3 1 4]
상위 3명 이름: ['하윤' '서연' '지호']
상위 3명 잔 수: [258 240 203]


<details>
<summary><b>기대 출력 보기 (문제 23)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
오름차순 위치: [2 0 4 1 3]
확인 - 그 위치의 값: [165 182 203 240 258]
상위 3명 위치: [3 1 4]
상위 3명 이름: [&#x27;하윤&#x27; &#x27;서연&#x27; &#x27;지호&#x27;]
상위 3명 잔 수: [258 240 203]
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 24. axis로 방향 정하기

지점 × 시간대 표에서 다음 두 가지를 구하세요.

1. **시간대별 합계** (아침 전체, 점심 전체, 저녁 전체, 야간 전체)
2. **지점별 합계** (강남 전체, 홍대 전체, 판교 전체)

각각 `axis=0`인지 `axis=1`인지 먼저 결정하세요.
판단 기준은 암기가 아니라 **결과가 몇 개 나와야 하는가**입니다.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [163]:
import numpy as np

branches = np.array(["강남", "홍대", "판교"])
timeslots = np.array(["아침", "점심", "저녁", "야간"])
sold = np.array([
    [42,  88, 61, 25],
    [30,  75, 94, 48],
    [55, 102, 47, 12],
])


# ===== 여기에 코드를 작성하세요 =====
print(f'표 \n{sold} \nshape: {sold.shape}')
print(f'Y축(컬럼)기준 = 시간대별 sum(axis=0) : {sold.sum(axis=0)}')
print(f'X축(행)기준 = 지점별 sum(axis=1) : {sold.sum(axis=1)}')
print("전체 합계(axis 없음):", sold.sum())

표 
[[ 42  88  61  25]
 [ 30  75  94  48]
 [ 55 102  47  12]] 
shape: (3, 4)
Y축(컬럼)기준 = 시간대별 sum(axis=0) : [127 265 202  85]
X축(행)기준 = 지점별 sum(axis=1) : [216 247 216]
전체 합계(axis 없음): 679


<details>
<summary><b>기대 출력 보기 (문제 24)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
표:
 [[ 42  88  61  25]
 [ 30  75  94  48]
 [ 55 102  47  12]] 
shape: (3, 4)
시간대: [&#x27;아침&#x27; &#x27;점심&#x27; &#x27;저녁&#x27; &#x27;야간&#x27;]
시간대별 합계(axis=0): [127 265 202  85] | 개수: 4
지점  : [&#x27;강남&#x27; &#x27;홍대&#x27; &#x27;판교&#x27;]
지점별 합계(axis=1): [216 247 216] | 개수: 3
전체 합계(axis 없음): 679
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Preview. AI로 이어지는 shape의 역할

암기 구간이 아닙니다. **계산 전에 shape를 확인하는 이유**만 가져갑니다.

### 문제 25. broadcasting - 다른 shape끼리 계산될 때와 안 될 때

`sold` 표(3행 4열)에 보너스 잔 수를 더하려고 합니다.

1. **시간대마다** 다른 보너스 `[1, 2, 3, 4]`를 더해 보세요. 될까요?
2. **지점마다** 다른 보너스 `[10, 20, 30]`을 그냥 더해 보세요. `try/except`로 감싸고 오류를 확인하세요.
3. 2번을 성공하도록 고치세요. (힌트: `reshape(3, 1)`)

각 단계에서 **shape를 먼저 출력**하고 예측한 뒤 실행하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [181]:
import numpy as np

sold = np.array([
    [42,  88, 61, 25],
    [30,  75, 94, 48],
    [55, 102, 47, 12],
])

# ===== 여기에 코드를 작성하세요 =====
print(f'sold표 \n{sold}')
print(f'sold shape  : {sold.shape}')
time_bonus = np.array([1, 2, 3, 4])
print(f'time_bonus 표 {time_bonus}')
print(f'time_bonus shape : {time_bonus.shape}')
try:
    apply_time_bonus = sold + time_bonus 
    #시간대별 보너스: (3,4)와 (4,) -> 마지막 축이 4로 일치하므로 성공

except ValueError as error:
    print("예상한 apply_time_bonus오류 종류:", type(error).__name__)
    print("이유: shape 형태.")
else:
    print(f' apply_time_bonus: \n{apply_time_bonus} ') 

branch_bonus = np.array([10, 20, 30])
print(f'branch_bonus \n{branch_bonus}')
print(f'branch_bonus.shape : {branch_bonus.shape}')
try:
    applied_branch_bonus = sold + branch_bonus 
    #지점별 보너스: (3,4)와 (3,) -> 마지막 축이 4 vs 3 -> 실패

except ValueError as error:
    print("예상한 branch_bonus 오류 종류:", type(error).__name__)
    

column_bonus = branch_bonus.reshape(3, 1)
print(f'branch_bonus.reshape \n{column_bonus} ')
print(f'branch_bonus.reshape : {column_bonus.shape}')
print("지점 보너스 적용:\n", sold + column_bonus)



sold표 
[[ 42  88  61  25]
 [ 30  75  94  48]
 [ 55 102  47  12]]
sold shape  : (3, 4)
time_bonus 표 [1 2 3 4]
time_bonus shape : (4,)
 apply_time_bonus: 
[[ 43  90  64  29]
 [ 31  77  97  52]
 [ 56 104  50  16]] 
branch_bonus 
[10 20 30]
branch_bonus.shape : (3,)
예상한 branch_bonus 오류 종류: ValueError
branch_bonus.reshape 
[[10]
 [20]
 [30]] 
branch_bonus.reshape : (3, 1)
지점 보너스 적용:
 [[ 52  98  71  35]
 [ 50  95 114  68]
 [ 85 132  77  42]]


<details>
<summary><b>기대 출력 보기 (문제 25)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
sold shape      : (3, 4)
time_bonus shape: (4,)
시간대 보너스 적용:
 [[ 43  90  64  29]
 [ 31  77  97  52]
 [ 56 104  50  16]]
예상한 오류 종류: ValueError
sold shape        : (3, 4)
branch_bonus shape: (3,)
고친 shape: (3, 1)
지점 보너스 적용:
 [[ 52  98  71  35]
 [ 50  95 114  68]
 [ 85 132  77  42]]
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

### 문제 26. 내적과 행렬 곱으로 매출·원가를 한 번에

두 단계로 진행합니다.

**(1) 벡터 내적**
메뉴 3개의 가격 `[4500, 5000, 5500]`과 판매 잔 수 `[10, 8, 6]`이 있습니다.
`*` 결과와 `np.dot()` 결과를 각각 출력하고, 무엇이 다른지 확인하세요.

**(2) 행렬 곱**
지점 2곳 × 메뉴 3개의 판매량 표와, 메뉴 3개 × [판매가, 원가] 표를 곱하세요.
결과가 무엇을 의미하는지 해석하고, 마지막으로 `A.T`(전치)로 행과 열이 바뀌는 것도 확인하세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [ ]:
import numpy as np

price_vec = np.array([4500, 5000, 5500])
cups_vec = np.array([10, 8, 6])

sales_count = np.array([  
    #아메리카노, 라떼, 콜드브루
    [10, 8, 6],   # 강남: 
    [12, 5, 9],   # 홍대
])
price_cost = np.array([
    #판매가, 원가
    [4500, 1200],   # 아메리카노: 
    [5000, 1500],   # 라떼
    [5500, 1800],   # 콜드브루
])

# ===== 여기에 코드를 작성하세요 =====
elementwise = price_vec * cups_vec
print("원소별 곱", elementwise ,"개수:", elementwise.size )

print("곱의 합", elementwise.sum() )
dot_value = np.dot(price_vec, cups_vec)
print("내적", dot_value , "개수:", dot_value.size)

# [내적] (2) (2,3) x (3,2) -> 안쪽 3이 서로 맞으므로 계산 가능, 결과는 (2,2)
result = np.dot(sales_count, price_cost)
print("지점별 [총매출, 총원가]:\n", result)
print("결과 shape       :", result.shape)
print("sales_count shape:", sales_count.shape)
print("price_cost  shape:", price_cost.shape)



원소별 곱 [45000 40000 33000] 개수: 3
곱의 합 118000
내적 118000 개수: 1
지점별 [총매출, 총원가]:
 [[118000  34800]
 [128500  38100]]
결과 shape       : (2, 2)
sales_count shape: (2, 3)
price_cost  shape: (3, 2)


<details>
<summary><b>기대 출력 보기 (문제 26)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
원소별 곱: [45000 40000 33000] | 개수: 3
곱의 합  : 118000
내적     : 118000 | 개수: 1
sales_count shape: (2, 3)
price_cost  shape: (3, 2)
결과 shape       : (2, 2)
지점별 [총매출, 총원가]:
 [[118000  34800]
 [128500  38100]]
강남 총매출 손계산: 118000
sales_count:
 [[10  8  6]
 [12  5  9]]
전치 결과:
 [[10 12]
 [ 8  5]
 [ 6  9]]
shape 변화: (2, 3) -&gt; (3, 2)
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)

---

## Final Challenge. 신메뉴 수익 분석

지금까지 배운 것을 한 문제에 모두 넣습니다.
답을 보기 전에 결과의 **개수와 첫 값**을 먼저 예측하세요.

### 문제 27. 신메뉴 4종 수익 분석 (종합)

신메뉴 4종의 판매가, 원가, 판매량 데이터입니다.

```python
new_menu   = np.array(["말차라떼", "흑임자라떼", "딸기크림", "밤라떼"])
sell_price = np.array([6000, 6500, 7000, 5800])
cost_price = np.array([2200, 2600, 3100, 2000])
sold_count = np.array([140, 95, 210, 60])
```

다음을 모두 구하세요.

1. 메뉴별 **개당 이익** (판매가 − 원가)
2. 메뉴별 **총이익** (개당 이익 × 판매량)
3. **전체 총이익**
4. 총이익이 **300,000원 이상**인 메뉴 이름
5. 총이익이 **높은 순서**의 메뉴 이름과 금액
6. 총이익이 가장 낮은 메뉴 이름 하나

**답을 쓰기 전에 예측하세요.**
- 4번의 결과는 몇 개가 나올 것 같습니까?
- 말차라떼의 총이익은 얼마입니까? (암산으로)

마지막에 "어떤 메뉴를 먼저 관리하겠는가"를 근거 한 문장으로 적으세요.

**[한글 설계] — 코드를 쓰기 전에 여기부터 채우세요**

```text
1.
2.
3.
```

In [ ]:
import numpy as np

new_menu = np.array(["말차라떼", "흑임자라떼", "딸기크림", "밤라떼"])
sell_price = np.array([6000, 6500, 7000, 5800])
cost_price = np.array([2200, 2600, 3100, 2000])
sold_count = np.array([140, 95, 210, 60])

# 계산 전에 shape가 모두 같은지 확인한다
print("메뉴 수:", new_menu.size)
print("shape 일치:",
      new_menu.shape == sell_price.shape == cost_price.shape == sold_count.shape)


# ===== 여기에 코드를 작성하세요 =====

<details>
<summary><b>기대 출력 보기 (문제 27)</b> &mdash; 먼저 스스로 실행한 뒤에 펼치세요</summary>

<pre>
메뉴 수: 4
shape 일치: True
메뉴        : [&#x27;말차라떼&#x27; &#x27;흑임자라떼&#x27; &#x27;딸기크림&#x27; &#x27;밤라떼&#x27;]
개당 이익   : [3800 3900 3900 3800]
메뉴별 총이익: [532000 370500 819000 228000]
전체 총이익 : 1949500
30만 이상   : [&#x27;말차라떼&#x27; &#x27;흑임자라떼&#x27; &#x27;딸기크림&#x27;] | 개수: 3
총이익 내림차순 메뉴: [&#x27;딸기크림&#x27; &#x27;말차라떼&#x27; &#x27;흑임자라떼&#x27; &#x27;밤라떼&#x27;]
총이익 내림차순 금액: [819000 532000 370500 228000]
가장 낮은 메뉴: 밤라떼 228000
</pre>
</details>

(아래는 정답 코드의 실행 결과입니다. 검증 셀 출력은 정답지에서 확인하세요.)